echo-pubsub/src/main/java/com/netflix/spinnaker/echo/pubsub/amazon/SQSSubscriber.java

```diff
@@ -0,0 +1,267 @@
+
+  private PubsubMessageHandler pubsubMessageHandler;
+  private MessageArtifactTranslator messageArtifactTranslator;
+
+  private NodeIdentity identity = new NodeIdentity();
+
+  Registry registry;
+
+  private final ARN queueARN;
+  private final ARN topicARN;
+
+  private String queueId = null;
+
+
+  public SQSSubscriber(ObjectMapper objectMapper,
+                       AmazonPubsubProperties.AmazonPubsubSubscription subscription,
+                       PubsubMessageHandler pubsubMessageHandler,
+                       AmazonSNS amazonSNS,
+                       AmazonSQS amazonSQS,
+                       Registry registry) {
+    this.objectMapper = objectMapper;
+    this.subscription = subscription;
+    this.pubsubMessageHandler = pubsubMessageHandler;
+    this.amazonSNS = amazonSNS;
+    this.amazonSQS = amazonSQS;
+    this.registry = registry;
+
+    this.messageArtifactTranslator = new MessageArtifactTranslator(subscription.readTemplatePath());
+    this.queueARN = new ARN(subscription.getQueueARN());
+    this.topicARN = new ARN(subscription.getTopicARN());
+  }
+
+  public String getWorkerName() {
+    return queueARN.getArn() + "/" + SQSSubscriber.class.getSimpleName();
+  }
+
+  @Override
+  public PubsubSystem pubsubSystem() {
+    return pubsubSystem;
+  }
+
+  @Override
+  public String subscriptionName() {
+    return subscription.getName();
+  }
+
+  @Override
+  public String getName() {
+    return getWorkerName();
+  }
+
+  @Override
+  public void run() {
+    log.info("Starting " + getWorkerName());
+
+    while (true) {
+      try {
+        listenForMessages();
+      } catch (Throwable e) {
+        log.error("Unexpected error running " + getWorkerName() + ", restarting", e);
+      }
+    }
+  }
+
+  private void listenForMessages() {
+    this.queueId = ensureQueueExists(
+      amazonSQS, queueARN, topicARN, subscription.getSqsMessageRetentionPeriodSeconds()
+    );
+    subscribeToTopic(amazonSNS, topicARN, queueARN);
+
+    while (true) {
+      ReceiveMessageResult receiveMessageResult = amazonSQS.receiveMessage(
+        new ReceiveMessageRequest(queueId)
+          .withMaxNumberOfMessages(AWS_MAX_NUMBER_OF_MESSAGES)
+          .withVisibilityTimeout(subscription.getVisibilityTimeout())
+          .withWaitTimeSeconds(subscription.getWaitTimeSeconds())
+      );
+
+      if (receiveMessageResult.getMessages().isEmpty()) {
+        log.debug("Received no messages");
+        continue;
+      }
+
+      receiveMessageResult.getMessages().forEach(message -> {
+        handleMessage(message);
+        registry.counter(getProcessedMetricId(subscriptionName())).increment();
+      });
+    }
+  }
+
+  private void handleMessage(Message message) {
+    String messageId = message.getMessageId();
+    String messagePayload = unmarshallMessageBody(message.getBody());
+
+    Map<String, String> stringifiedMessageAttributes = message.getMessageAttributes().entrySet().stream()
```

Comment:
Curious why we went with `Map<String, String>` vs `Map<String, Object>` @jtk54?

Reply:
I assume because we're just putting them in the pipeline context as strings from here.